File: testing/matrix/benchmark.py
import json, time, argparse
from algo_matrix import create_schedules
import numpy as np

matrix_shape = (12*31*24*60)
packed_64_shape = (8370)
packed_32_shape = (16740)

parser = argparse.ArgumentParser()
parser.add_argument('--courses', dest='course_list', help='List of courses to create schedules from',nargs='+')
parser.add_argument('--term', dest='term', help='Term code')
parser.add_argument('--log',dest='log',help='Log results to file',default=False,action='store_true')
args = parser.parse_args()

file = f'table-{args.term}.json'
with open(file,'r') as file:
    data = json.load(file)

courses = [course for course in data if course['name'] in args.course_list]

start_total = time.time()
for course in courses:
    for group in course['groups']:
        if len(group['matrix']) != 0:
            group['matrix'] = np.array(group['matrix'],dtype=np.uint32).view(np.uint64)

start = time.time()
results = create_schedules(courses)
end = time.time()
duration = round(end - start, 2)
total_duration = round(end - start_total,2)

print(f'algorithm duration: {duration} seconds')
print(f'total duration: {total_duration} seconds')
print(f'count: {len(results)}')

if args.log:
    results = [[group['crns'] for group in result] for result in results]
    results.insert(0, {'algorithm': 'matrix', 'duration': str(duration) + ' seconds', 'count': len(results)})
    with open(f'results-{args.term}.json','w') as file:
        json.dump(results, file, indent=2)

File: testing/hash_grouping/go/benchmark.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"time"
	"flag"
	"strings"
)

func filter(arr []Course, valid []string) []Course {
	result := []Course{}
	for i := range arr {
		for j := range valid {
			if arr[i].Name == valid[j] {
				result = append(result, arr[i])
			}
		}
	}
	return result
}

func main() {
	term := flag.Int("term", -1,"Term to create schedules for")
	names := flag.String("courses", "", "List of courses to create schedules from")
	flag.Parse()
	course_names := strings.Split(*names, ",")
	path := fmt.Sprintf("table-%v.json", *term)

	content, err := ioutil.ReadFile(path)
	if err != nil {
		log.Fatal(err)
	}

	var data []Course
	err = json.Unmarshal(content, &data)
	if err != nil {
		log.Fatal(err)
	}
	
	courses := filter(data, course_names)

	start := time.Now()
	results := CreateSchedules(courses)
	duration := time.Since(start).Seconds()

	fmt.Printf("duration: %.2f seconds\n", duration)
	fmt.Println("count: ", len(results))
}
